pymc_experimental/utils/cache.py

```diff
@@ -0,0 +1,188 @@
+import hashlib
+import os
+import sys
+from typing import Callable, Literal
+
+import arviz as az
+import numpy as np
+import pymc
+import pytensor
+from pymc import (
+    modelcontext,
+    sample,
+    sample_posterior_predictive,
+    sample_prior_predictive,
+)
+from pymc.model.fgraph import fgraph_from_model
+from pytensor.compile import SharedVariable
+from pytensor.graph import Constant, FunctionGraph, Variable
+from pytensor.scalar import ScalarType
+from pytensor.tensor import TensorType
+from pytensor.tensor.random.type import RandomType
+from pytensor.tensor.type_other import NoneTypeT
+
+import pymc_experimental
+
+
+def hash_data(c: Variable) -> str:
+    if isinstance(c.type, NoneTypeT):
+        return "None"
+    if isinstance(c.type, (ScalarType, TensorType)):
+        if isinstance(c, Constant):
+            arr = c.data
+        elif isinstance(c, SharedVariable):
+            arr = c.get_value(borrow=True)
+        arr_data = arr.view(np.uint8) if arr.size > 1 else arr.tobytes()
+        return hashlib.sha1(arr_data).hexdigest()
+    else:
+        raise NotImplementedError(f"Hashing not implemented for type {c.type}")
+
+
+def get_name_and_props(obj):
+    name = str(obj)
+    props = str(getattr(obj, "_props", lambda: {})())
```

Comment:
```suggestion
    name = str(obj)
    if hasattr(obj, "_props"):
        prop_dict = obj._prop_dict()
        props = str(
            {k: get_name_and_props(v) for k, v in prop_dict.items()}
        )
    else:
        props = str({})
    
    name = str(obj)
    props = str(getattr(obj, "_props", lambda: {})())
```

Reply:
This is just to make sure that potential recursion of _props is handled.

Reply:
_props are not recursive